library/std/src/os/windows/net/addr.rs

```diff
@@ -0,0 +1,83 @@
+#![unstable(feature = "windows_unix_domain_sockets", issue = "56533")]
+
+use crate::os::raw::{c_char, c_int};
+use crate::path::Path;
+use crate::sys::c::{self, SOCKADDR};
+use crate::sys::cvt;
+use crate::{io, mem};
+pub fn sockaddr_un(path: &Path) -> io::Result<(c::sockaddr_un, c_int)> {
+    let mut addr: c::sockaddr_un = unsafe { mem::zeroed() };
+    addr.sun_family = c::AF_UNIX;
+    // Winsock2 expects 'sun_path' to be a Win32 UTF-8 file system path
+    let bytes = path
+        .to_str()
+        .map(|s| s.as_bytes())
+        .ok_or(io::Error::new(io::ErrorKind::InvalidInput, "path contains invalid characters"))?;
+
+    if bytes.contains(&0) {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "paths may not contain interior null bytes",
+        ));
+    }
+
+    if bytes.len() >= addr.sun_path.len() {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "path must be shorter than SUN_LEN",
+        ));
+    }
+    for (dst, src) in addr.sun_path.iter_mut().zip(bytes.iter()) {
+        *dst = *src as c_char;
+    }
+    // null byte for pathname addresses is already there because we zeroed the
+    // struct
+
+    let mut len = sun_path_offset(&addr) + bytes.len();
+    match bytes.first() {
+        Some(&0) | None => {}
+        Some(_) => len += 1,
+    }
+    Ok((addr, len as _))
+}
+fn sun_path_offset(addr: &c::sockaddr_un) -> usize {
+    // Work with an actual instance of the type since using a null pointer is UB
+    let base = addr as *const _ as usize;
+    let path = &addr.sun_path as *const _ as usize;
+    path - base
```

Comment:
Please use https://doc.rust-lang.org/std/mem/macro.offset_of.html.

```suggestion
    offset_of!(c::sockaddr_un, sun_path)
```